### PR TITLE
feat: [CDS-80874]: create K8s release metadata util

### PR DESCRIPTION
**Pull request by abhinavsingh1196**
## Describe your changes
Utility to check if k8s releases conflict, create a log-able message for Harness UI. Can be used by respective k8s delegate task handlers later

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54055)

### CODE REVIEW

File: 960-api-services/src/main/java/io/harness/k8s/K8sReleaseDiffCalculator.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    return sb.toString();
+  }
+
+  public ReleaseMetadata getPreviousReleaseMetadata(IK8sReleaseHistory releaseHistory, boolean inCanaryWorkflow) {
+    Optional<IK8sRelease> previousReleaseOptional = getPreviousRelease(releaseHistory, inCanaryWorkflow);
+    return previousReleaseOptional.map(IK8sRelease::getReleaseMetadata).orElse(null);
+  }
+
+  private Optional<IK8sRelease> getPreviousRelease(IK8sReleaseHistory releaseHistory, boolean inCanaryWorkflow) {
+    if (inCanaryWorkflow) {
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/K8sReleaseDiffCalculator.java:61]: Is this logic correct? Usually we're doing in reverse

Comment (abhinavsingh1196) [reply 960-api-services/src/main/java/io/harness/k8s/K8sReleaseDiffCalculator.java:61]: Thought of it this way: 

when in canary workflow:
- during canary deploy:
  1. get release history (r1-SUCCESS, r2-SUCCESS, r3-FAILED)
  2. create release r4, with metadata m4
  3. `getPreviousRelease` would return r2, so we will compare metadata m4 with m2
- during rolling deploy:
  1. get release history (r1-SUCCESS, r2-SUCCESS, r4-INPROGRESS)
  2. create release r4, with metadata m4
  3. `getPreviousRelease` would return r2, so we will compare metadata m4 with m2 


when in rolling flow:
- get release history (r1-SUCCESS, r2-SUCCESS, r3-FAILED)
- create release r4, with metadata m4
- `getPreviousRelease` would return r3, so we will compare metadata m3 with m2

I wasn't sure how to keep the behaviour exactly. I could make this consistent if you prefer that, i.e. either use latest release always OR use last successful release always. 
I guess if we use last successful release then we won't have to handle these `inCanaryWorkflow` cases 🤔

Comment (abosii-harness) [reply 960-api-services/src/main/java/io/harness/k8s/K8sReleaseDiffCalculator.java:61]: yeah, I believe we should just use getLatestRelease

**Review by abosii-harness — DISMISSED** [review]

**Comment by abhinavsingh1196** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

**Review by alebuh — APPROVED** [review]

**Comment by abhinavsingh1196** [comment]
trigger ut7

**Comment by abhinavsingh1196** [comment]
trigger ut7

**Review by markoharness — APPROVED** [review]

**Pull request merged**